### ISSUE TITLE
Flow fails to check Disjoint Unions as object value type

### ISSUE THREAD

**Issue by AJIh** [facebook/flow#1959]
```javascript
type ASTLiteralNode = {
  type: 'Literal',
  value: any
};

type ASTThisNode = {
  type: 'ThisExpression'
};

class AST {
  static Literal = 'Literal';
  static ThisExpression = 'ThisExpression';

  static constants: { [key: string]: (ASTThisNode | ASTLiteralNode) } = {
    'this': { type: AST.ThisExpression },
    'null': { type: AST.Literal, value: null },
    'true': { type: AST.Literal, value: true },
    'false': { type: AST.Literal, value: false }
  };
```

I got the errors: 

```
src/parse.js:242
242:     'null': { type: AST.Literal, value: null },
                         ^^^^^^^^^^^ string. Expected string literal `ThisExpression`, got `Literal` instead
228:   type: 'ThisExpression'
             ^^^^^^^^^^^^^^^^ string literal `ThisExpression`

src/parse.js:243
243:     'true': { type: AST.Literal, value: true },
                         ^^^^^^^^^^^ string. Expected string literal `ThisExpression`, got `Literal` instead
228:   type: 'ThisExpression'
             ^^^^^^^^^^^^^^^^ string literal `ThisExpression`

src/parse.js:244
244:     'false': { type: AST.Literal, value: false }
                          ^^^^^^^^^^^ string. Expected string literal `ThisExpression`, got `Literal` instead
228:   type: 'ThisExpression'
             ^^^^^^^^^^^^^^^^ string literal `ThisExpression`
```

**Comment by AJIh** [comment]
Using literal type solves the problem.

```javascript
  static constants: { [key: string]: (ASTThisExpressionNode | ASTLiteralNode) } = {
    'this': { type: 'ThisExpression' },
    'null': { type: 'Literal', value: null },
    'true': { type: 'Literal', value: true },
    'false': { type: 'Literal', value: false }
  };
```

But flow works perfectly if I just use `AST.Literal`.

```javascript
  static constants: { [key: string]: ASTLiteralNode } = {
    'null': { type: AST.Literal, value: null },
    'true': { type: AST.Literal, value: true },
    'false': { type: AST.Literal, value: false }
  };
```

Is it a bug? Or I missed something in the doc?😳

**Comment by patrys** [comment]
This is likely part of a larger problem with disjoint unions.

```
type A = {type: '1', value: B} | {type: '2': value: C}
```

The above is treated as: `{type: '1', value: B & C} | {type: '2', value: B & C}`. Key tags that exist in multiple disjoint alternatives are matched against each one of their types.

**Comment by masylum** [comment]
Is there any workaround to this problem? I'm getting the same problem with this types:

```js
type BaseAction = { icon: string }
type ActionWithUploads = BaseAction & { onDrop: (files: Array<File>) => void }
type ActionWithLink = BaseAction & { href: string }
type Action = (ActionWithLink) | (ActionWithUploads)
```

Flow is failing to accept an object of the type `Action` that does not contain both `onDrop` and `href`.

Thanks in advance

**Comment by masylum** [comment]
Also, it wouldn't hurt in the meantime to mention this issue in the documentation. Many people may be trying to do this and thinking they are doing something wrong.

**Comment by LoganBarnett** [comment]
@masylum, I tried dropping your snippet above into the [try flow editor](https://flowtype.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgBCAhgM54CCAxhgJZwB2YAvGAN5i1WMBcYpGAE60GAczABfTLgLU6jAOq0MACwCqOeMQAmpFkTKUa9JgDJ2YRgBFBcHHwAUUWjDyk+FQYOJYAPADEXPAA+AEoWYLAANzhabUlpfDA5EyVVABkRAGt9EnIUxjBzDhVBPCg+AWExBOwkgqZWBwa0lUyGLPCAHzBm40VldU04HVJQ9CA) and it showed no problems. Do you have an example of using these types where it breaks?

**Comment by patrys** [comment]
It used to fail with simple test cases like this but it seems to work fine now:

```javascript
type A = {type: '1'; value: string} | {type: '2'; value: number}

function foo(x: A) {
  switch (x.type) {
    case '1': return x.value.toUpperCase()
    case '2': return x.value * 2
  }
}
```

**Comment by frontsideair** [comment]
I have built a [failing case](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVBLAtgBzgJwBcwAlAUwEMBjQgGjAG8wBhOXOAOzI+IF8wo+NmADk+SjRHpCATxxkwAQTABeRmEIUA5gC4wHAK4wY9AG4UYBsnsPGwvVLPlgAQqvWbdoigBMfGDi0RMwsrPQBnQnwArXtHOQVFNzVlAB9XdCpOSLAKPWU1Jk8bIxMwc0trfVK41CoYCnDwljAyAA9Cbh9m1nYuHgAeCg4ZeiSAPkZUMDBIik73CmmwcQ4fMnwwAAoASimZmYGAIwNCQk4wTmYYDCoAaxUGXdVJwgALDHCAOnCyQgBlTSdLYUHb8YDjZYOXhAA) and I couldn't find a workaround. Can anyone check?

**Comment by asolove** [comment]
I can simplify that case by removing all the React bits:

```js
type A = { tag: 'stuff', value: null }
type B = { tag: 'adding', value: string }
type AB = A | B

const a: A = { tag: 'stuff', value: null }

let b: AB = a; // Type-checks, as expected
let c: $Shape<AB> = a; // Type error, but should work?
```

So I think it's specifically an issue with unions and `$Shape`.

**Comment by frontsideair** [comment]
I get that it's a complicated case. But what should be done to support my use case? I think it's a valid use case and not an edge case.

I got a tip on Twitter suggesting I wrap my types in another object, and it worked, but it's a lot of hassle for a case that should just work.

Here's the tip: https://mobile.twitter.com/calebmer/status/901067504537268224

**Comment by asolove** [comment]
The reason why this is complicated is because of the merging together of a current state and some changes to it in a context that is a union type. The extra object makes React `setState` work because that nested object fully replaces the existing one rather than being merged into it.

If you try to think about the type of the merge operation, I think it makes sense that this isn't doable statically when Flow checks.

For example, let's take this context of a type on a variable:

```js
declare var x: { tag: 'a'  } | { tag: 'b', val: number };
```

If we merge a new object into `x`, what is the type of object that is allowed to be merged into `x`?

```js
x = {...x, { val: 2 }}; // is this allowed?
x = {...x, { val: 'b' }} // what about this?
```

Without knowing which of the cases (a or b), `x` currently is, we can't know what type of object is allowed to be merged in. For example. the first line only type checks if we know the value of `x` is in the second case (has tag `b`) and the second line is only allowed if the object is already in the first case (has tag `a`). But flow doesn't know which state the object will be in at runtime, so it simply can't allow either of these to type-check.

So, hopefully that explains why `$Shape` has trouble with unions for valid reasons, and also why the nested object isn't just a syntax trick, but actually makes the logic correct.

**Comment by frontsideair** [comment]
Yeah, that makes a lot of sense. The tool I reached first was in fact `replaceState`, but it was [deprecated](https://github.com/facebook/react/issues/3236) sadly. Looks like I'll have to do it with a wrapped object for now.
